business/namespaces.go

```diff
@@ -232,140 +144,46 @@ func (in *NamespaceService) GetNamespaces(ctx context.Context) ([]models.Namespa
-		selectors := make([]labels.Selector, 0)
-		for _, selector := range discoverySelectors {
-			ls, err := meta_v1.LabelSelectorAsSelector(selector)
-			if err != nil {
-				return nil, fmt.Errorf("error initializing discovery selectors filter, invalid discovery selector: %v", err)
-			}
-			selectors = append(selectors, ls)
-		}
-
-		// 2. range over all namespaces to get discovery namespaces, notice each selector result is ORed (as per Istio convention)
-		selectedNamespaces := make([]models.Namespace, 0)
-		for _, ns := range resultns {
-			if ns.Name == in.conf.IstioNamespace {
-				selectedNamespaces = append(selectedNamespaces, ns) // we always want to return the control plane namespace
-			} else {
-				for _, selector := range selectors {
-					if selector.Matches(labels.Set(ns.Labels)) {
-						selectedNamespaces = append(selectedNamespaces, ns)
-						break
-					}
-				}
-			}
-		}
-		namespaces = selectedNamespaces
-		resultns = namespaces
-	}
-
-	// exclude namespaces that are:
-	// 1. to be filtered out via the exclude list
-	// 2. to be filtered out via the label selector
-	// Note that the control plane namespace is never excluded
-	excludes := in.conf.API.Namespaces.Exclude
-	if len(excludes) > 0 || labelSelectorExclude != "" {
-		resultns = []models.Namespace{}
-	NAMESPACES:
-		for _, namespace := range namespaces {
-			if namespace.Name != in.conf.IstioNamespace {
-				if len(excludes) > 0 {
-					for _, excludePattern := range excludes {
-						if match, _ := regexp.MatchString(excludePattern, namespace.Name); match {
-							continue NAMESPACES
-						}
-					}
-				}
-				if labelSelectorExclude != "" {
-					if namespace.Labels[labelSelectorExcludeName] == labelSelectorExcludeValue {
-						continue NAMESPACES
-					}
-				}
-			}
-			resultns = append(resultns, namespace)
-		}
-	}
-
 	// store only the filtered set of namespaces in cache for the token
 	namespacesPerCluster := make(map[string][]models.Namespace)
-	for _, ns := range resultns {
+	for _, ns := range namespaces {
 		namespacesPerCluster[ns.Cluster] = append(namespacesPerCluster[ns.Cluster], ns)
 	}
 	for cluster, ns := range namespacesPerCluster {
 		in.kialiCache.SetNamespaces(in.userClients[cluster].GetToken(), ns)
 	}
 
-	return resultns, nil
+	return namespaces, nil
 }
 
+// getNamespacesByCluster returns the namespaces for the given cluster. The namespaces are filtered such
+// that only those namespaces that match discovery selectors are returned.
 func (in *NamespaceService) getNamespacesByCluster(ctx context.Context, cluster string) ([]models.Namespace, error) {
-	configObject := in.conf
-
-	labelSelectorInclude := configObject.API.Namespaces.LabelSelectorInclude
 
 	var namespaces []models.Namespace
-	_, queryAllNamespaces := in.isAccessibleNamespaces["**"]
+
+	mesh, ok := in.kialiCache.GetMesh()
+	if !ok {
+		log.Debugf("Cannot obtain mesh details for cluster [%v]; will not be able to filter on Istio discovery selectors if they exist", cluster)
+	}
```

Comment:
> We need to fail at some point if there's no discovery selectors configured for kiali and we can't autodiscover the istio discovery selectors.

Only if `cluster_wide_access` == `false`, no? I don't think we want to force/require users to define discovery selectors if Kiali is given cluster-wide access (even [Istio's docs](https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig) say that if Istio discovery selectors aren't defined, all namespaces are processed: `If omitted, Istio will use the default behavior of processing all namespaces in the cluster.`)

Reply:
Well since they are used as a form of access control / scoping for Kiali, we don't want Kiali to suddenly start showing namespaces outside of the istio discovery selectors because Kiali can't get the list of discovery selectors from istio for whatever reason.

I'm wondering if we drop the autodiscovery of the istio discovery selectors altogether and default the kiali discovery selectors in the config to process all namespaces?

Reply:
That's a possibility. IIRC, the Istio Discovery Selectors isn't really technically defining what's considered in the mesh anyway.. I thought that was just a way for Istio to limit what it looks for to, say, process auto injection. But it is possible to manually inject sidecars in namespaces that aren't in the discovery selectors and that would work (in other words, DS in Istio was just a feature to increase performance; it wasn't a feature to literally define the mesh namespaces).

Let's think about this some more. This would make Kiali simpler because it would mean both server and operator do not have to worry about auto-discoverying the one (or more) istiod control planes to determine what namespaces to look at. If the user doesn't want Kiali to make all-namespaces visible, they would need to explicitly configure Kiali in that case, rather than rely on Istio DS as a backup. Off the top of my head, I'm thinking that may not be a bad compromise.

Reply:
I just reviewed the semantics of Istio discovery selectors, and those do seem to be the way users can control which namespaces Istio considers "in the mesh" ... see blog: https://istio.io/v1.13/blog/2021/discovery-selectors/

Note they make the same distinction we do with our "cluster_wide_access=true" : `Note: discoverySelectors is not a security boundary. Istiod will continue to have access to all namespaces even when you have configured your discoverySelectors.` The difference is we provide `cluster_wide_access=false` for our paranoid users.

Anyway - what should we do here? Right now, if there are NO `deployment.discovery_selectors.default` selectors defined in the Kiali CR, AND we get an error trying to obtain the control plane's discovery selectors (or we just can't get them for whatever reason), we keep going. We should probably fail.

But there are some other things we need to be concerned about if we are to use Istio discovery selectors:
1. What happens if we have multiple control planes (think: canary upgrades of Istio)?
2. What happens if there is an external control plane(s) in remote clusters and it/they have Istio discovery selectors? We don't attempt to look for those at all.

I'm starting to think we need to rip out anything to do with Kiali attempting to find and use Istio discovery selectors and we need to document that we do not use them. If the Istio service mesh has discovery selectors, the Kiali admin needs to configure Kiali with the same discovery selectors. I don't think we can effectively auto-discovery them for all the different use cases (multiple control planes, external control planes, istiod that isn't accessible to Kiali or we get an error trying to get these istio selectors).